### PR TITLE
ubootstrategy: add support for initial state

### PR DESCRIPTION
**Pull request by nlabriet**
**Description**

Update the uboot strategy to feature initial state, usable from pytest and labgrid-client.
Unnecessary actions to reach the requested state are not performed.

Tested with labgrid-client in my setup, transitioning:
- from off to uboot with initial state: off
- from uboot to shell with initial state: uboot

**Checklist**
- [ ] Documentation for the feature
Documentation already present
- [X] PR has been tested
Manual tests in my setup
- [ ] Man pages have been regenerated
No man page impacted, this can serve as an example to implement a strategy with initial state support

### CODE REVIEW

**Review by Bastian-Krause — CHANGES_REQUESTED** [review]
The driver activation should happen inside the `force()` method, not in `transition()`. See `labgrid/strategy/bareboxstrategy.py` or `labgrid/strategy/shellstrategy.py`.

**Comment by nlabriet** [comment]
Right, I've changed the force method to include the driver activations.

File: labgrid/strategy/ubootstrategy.py [hunk 1]
```diff
@@ -41,19 +41,35 @@ def transition(self, status):
             self.target.activate(self.power)
             self.power.off()
         elif status == Status.uboot:
-            self.transition(Status.off)
+            if self.status != Status.off:
+                self.transition(Status.off)
```
Comment (Bastian-Krause) [inline labgrid/strategy/ubootstrategy.py:45]: This is not needed. The transition method does nothing if the requested status is the same as the saved `self.status`, see above ("nothing to do"). Think of `transition()` as a state machine.

File: labgrid/strategy/ubootstrategy.py [hunk 2]
```diff
@@ -41,19 +41,35 @@ def transition(self, status):
...
             # cycle power
             self.power.cycle()
             # interrupt uboot
             self.target.activate(self.uboot)
         elif status == Status.shell:
-            # transition to uboot
-            self.transition(Status.uboot)
+            if self.status != Status.uboot:
+                # transition to uboot
+                self.transition(Status.uboot)
```
Comment (Bastian-Krause) [inline labgrid/strategy/ubootstrategy.py:54]: Same as above.

**Review by Bastian-Krause — CHANGES_REQUESTED** [review]

File: labgrid/strategy/ubootstrategy.py (hunk 1 above)
Comment (nlabriet) [reply labgrid/strategy/ubootstrategy.py]: right it's called recursively

File: labgrid/strategy/ubootstrategy.py (hunk 2 above)
Comment (nlabriet) [reply labgrid/strategy/ubootstrategy.py]: ok, reverting to original

File: labgrid/strategy/ubootstrategy.py [hunk 3]
```diff
@@ -57,3 +57,17 @@ def transition(self, status):
...
             raise StrategyError(f"no transition found from {self.status} to {status}")
         self.status = status
+
+    def force(self, status):
+        if not isinstance(status, Status):
+            status = Status[status]
+        if status == Status.off:
+            self.target.activate(self.power)
+        elif status == Status.uboot:
+            self.target.activate(self.console)
```
Comment (Bastian-Krause) [inline labgrid/strategy/ubootstrategy.py:67]: You can drop this line. The activation of the UBootDriver (which binds against the ConsoleProtocol) in the next line activates it anyway. Activating the ConsoleProtocol explicitly is only needed if we need to to it early, i.e. before turning power on, so we don't miss early characters on the serial interface.

**Review by Bastian-Krause — COMMENTED** [review]
Sorry, this one slipped through earlier.

File: labgrid/strategy/ubootstrategy.py (hunk 3 above)
Comment (nlabriet) [reply labgrid/strategy/ubootstrategy.py]: Line removed. I tested the transition form initial uboot state to shell state, it works!

**Review by Bastian-Krause — APPROVED** [review]

**Pull request merged**